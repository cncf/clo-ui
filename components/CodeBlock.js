module.exports = require('../dist/cjs/components/CodeBlock.js');
