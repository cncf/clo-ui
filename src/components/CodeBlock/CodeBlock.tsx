import { isUndefined } from 'lodash';
import React from 'react';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { docco } from 'react-syntax-highlighter/dist/cjs/styles/hljs';
import { tomorrowNight } from 'react-syntax-highlighter/dist/cjs/styles/hljs';

import { ButtonCopyToClipboard } from '../ButtonCopyToClipboard/ButtonCopyToClipboard';
import styles from './CodeBlock.module.css';

export interface ICodeBlockProps {
  language: string;
  content: string;
  withCopyBtn: boolean;
  label?: string;
  darkCode?: boolean;
  style?: any;
  effective_theme: string;
}

export const CodeBlock = (props: ICodeBlockProps) => {
  const isDarkActive = (props.effective_theme === 'dark' && isUndefined(props.darkCode)) || props.darkCode;

  return (
    <div data-testid="code" className={`d-flex flex-row align-items-center pb-2 ${styles.codeBlock}`}>
      <SyntaxHighlighter
        language={props.language}
        style={isDarkActive ? tomorrowNight : docco}
        customStyle={{
          backgroundColor: 'var(--bg-code)',
          color: 'var(--color-font)',
          padding: '1rem',
          marginBottom: 0,
          width: props.withCopyBtn ? 'calc(100% - 1rem - 32px)' : '100%',
          ...props.style,
        }}
      >
        {props.content}
      </SyntaxHighlighter>

      {props.withCopyBtn && (
        <ButtonCopyToClipboard
          text={props.content}
          label={props.label || 'Copy code to clipboard'}
          wrapperClassName="ms-3"
        />
      )}
    </div>
  );
};
