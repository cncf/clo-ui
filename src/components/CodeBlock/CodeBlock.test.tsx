import { act, render, screen } from '@testing-library/react';

import { CodeBlock } from './CodeBlock';

const defaultProps = {
  language: 'markdown',
  content: '##Sample',
  label: 'Copy btn',
  withCopyBtn: true,
  effective_theme: 'light',
};

const flushPromises = () => new Promise((resolve) => setTimeout(resolve, 0));

describe('CodeBlock', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', async () => {
    const { asFragment } = render(<CodeBlock {...defaultProps} />);
    await act(async () => {
      await flushPromises();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', async () => {
    render(<CodeBlock {...defaultProps} />);
    await act(async () => {
      await flushPromises();
    });

    const code = screen.getByTestId('code');
    expect(code).toBeInTheDocument();
    expect(code).toHaveTextContent('##Sample');

    expect(screen.getByRole('button', { name: 'Copy btn' })).toBeInTheDocument();
  });
});
