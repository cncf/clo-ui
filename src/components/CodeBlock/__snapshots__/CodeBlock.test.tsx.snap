// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CodeBlock creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-row align-items-center pb-2 codeBlock"
    data-testid="code"
  >
    <pre
      class="mb-0 flex-grow-1"
      style="display: block; overflow-x: auto; padding: 1rem; color: var(--color-font); background-color: var(--bg-code); margin-bottom: 0px; flex: 1 1 auto; white-space: pre-wrap; word-break: break-word; overscroll-behavior: contain; width: calc(100% + (-1 * (32px + 1rem)));"
    >
      <code
        class="language-markdown"
        style="white-space: pre;"
      >
        <span
          style="color: rgb(25, 70, 157);"
        >
          ##Sample
        </span>
      </code>
    </pre>
    <div
      class="position-relative ms-3"
    >
      <button
        aria-label="Copy btn"
        class="btn btn-sm btn-primary rounded-0 btn"
        type="button"
      >
        <div
          aria-hidden="true"
          class="d-flex flex-row align-items-center"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="13"
              rx="2"
              ry="2"
              width="13"
              x="9"
              y="9"
            />
            <path
              d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
